cpp/src/binaryop/binaryop.cpp

```diff
@@ -17,293 +17,307 @@
+    // before checking for equality because the same category values can mean
+    // different things for different columns
+
+    // make temporary columns which will have synced categories
+    auto temp_lhs = cudf::allocate_like(*lhs);
+    auto temp_rhs = cudf::allocate_like(*rhs);
+    gdf_column* input_cols[2] = {lhs, rhs};
+    gdf_column* temp_cols[2] = {&temp_lhs, &temp_rhs};
+
+    // sync categories
+    sync_column_categories(input_cols, temp_cols, 2);
+
+    // now it's ok to directly compare the column data
+    auto err =
+        binops::compiled::binary_operation(out, &temp_lhs, &temp_rhs, ope);
     if (err == GDF_UNSUPPORTED_DTYPE || err == GDF_INVALID_API_CALL)
-        binops::jit::binary_operation(out, lhs, rhs, ope);
+      binops::jit::binary_operation(out, &temp_lhs, &temp_rhs, ope);
+
+    // TODO: Need a better way to deallocate temporary columns
+    RMM_TRY(RMM_FREE(temp_lhs.data, 0));
+    RMM_TRY(RMM_FREE(temp_rhs.data, 0));
+    if (temp_lhs.valid != nullptr) RMM_TRY(RMM_FREE(temp_lhs.valid, 0));
+    if (temp_rhs.valid != nullptr) RMM_TRY(RMM_FREE(temp_rhs.valid, 0));
+    NVCategory::destroy(
+        reinterpret_cast<NVCategory*>(temp_lhs.dtype_info.category));
+    NVCategory::destroy(
+        reinterpret_cast<NVCategory*>(temp_rhs.dtype_info.category));
+    return;
+  }
+  auto err = binops::compiled::binary_operation(out, lhs, rhs, ope);
+  if (err == GDF_UNSUPPORTED_DTYPE || err == GDF_INVALID_API_CALL)
+    binops::jit::binary_operation(out, lhs, rhs, ope);
 }
 
-void binary_operation(gdf_column* out, gdf_column* lhs, gdf_column* rhs, const std::string& ptx) {
-    // Check for null pointers in input
-    CUDF_EXPECTS((out != nullptr) && (lhs != nullptr) && (rhs != nullptr),
-        "Input pointers are null");
-
-    // Check for 0 sized data
-    if((out->size == 0) && (lhs->size == 0) && (rhs->size == 0)) return;
-    CUDF_EXPECTS((out->size == lhs->size) && (lhs->size == rhs->size),
-        "Column sizes don't match");
-
-    // Check for null data pointer
-    CUDF_EXPECTS((out->data != nullptr) &&
-                 (lhs->data != nullptr) &&
-                 (rhs->data != nullptr) , 
-        "Column data pointers are null");
-
-    // Check for datatype
-    CUDF_EXPECTS((out->dtype == lhs->dtype) && (lhs->dtype == rhs->dtype) &&
-                 (out->dtype == GDF_FLOAT32 || out->dtype == GDF_FLOAT64 || 
-                  out->dtype == GDF_INT64   || out->dtype == GDF_INT32     ),
-        "Invalid/Unsupported datatype");
-    
-    binops::binary_valid_mask_and(out->null_count, out->valid, lhs->valid, rhs->valid, rhs->size);
-
-    binops::jit::binary_operation(out, lhs, rhs, ptx);
+gdf_column binary_operation(const gdf_column& lhs, const gdf_column& rhs,
+                            const std::string& ptx, gdf_dtype output_type) {
+  CUDF_EXPECTS((lhs.size == rhs.size), "Column sizes don't match");
+
+  gdf_column output;
```

Comment:
Need to make sure everything is zero-initialized. 
```suggestion
  gdf_column output{};
```